Feature request: neat Scala 3 definitions and calls – keep parens on the same line as arguments

- **Version**: 3.0.0-RC4
- **Integration**: sbt-scalafmt 2.4.2
- **Configuration**:
```
version = 3.0.0-RC4
maxColumn = 40

runner.dialect = scala3

danglingParentheses {
  defnSite = false
}
```

## Steps

Given code like this:
```scala
def notAnInterestingFunction
    (first: String)
    (second: String, third: String)
    (using fourth: Int)
    : (String, String, String, String) =
  (
    "The quick brown fox",
    "jumped over the lazy",
    "dog.",
    "Maybe."
  )

def f = 
  notAnInterestingFunction
    ("first")
    ("second", "third")
    (using 52)
```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

Scalafmt formats code like this, which is not an obvious improvement in readability:
```scala
def notAnInterestingFunction(
    first: String)(
    second: String,
    third: String)(using fourth: Int)
    : (String, String, String, String) =
  (
    "The quick brown fox",
    "jumped over the lazy",
    "dog.",
    "Maybe."
  )

def f =
  notAnInterestingFunction("first")(
    "second",
    "third"
  )(using 52)
```

## Expectation

I should like the formatted output to look like the original code, with:

1. each group of arguments on its own line;
2. with the parens around each argument group appearing on the same line as those arguments (provided each group of parens/arguments fits on a single line).

## Notes

1. This issue a write-up of my comment at https://github.com/scalameta/scalafmt/issues/2522#issuecomment-845961183 and the one following.

2. It may be helpful to able to control the formatting of definitions and calls independently.

@kitbellew I’m really sorry to bother you again, but I may have found a parser issue.

With this configuration:

```
version = "3.0.0-RC4-SNAPSHOT"
maxColumn = 40

runner.dialect = scala3

indent {
  main = 3
  significant = 6
  callSite = 6
  defnSite = 6
}

danglingParentheses {
  defnSite = false
  callSite = false
}

newlines {
  beforeOpenParenDefnSite = true
  beforeOpenParenCallSite = true
  sometimesBeforeColonInMethodReturnType = true
  source=fold
  beforeMultiline = unfold
  forceBeforeMultilineAssign = any
}
```

This code:

```scala
package x.y

class CompositeThing(components: String*)

object ExampleThing
    extends CompositeThing("One", "Two", "Three", "Four")
```

Is formatted to:

```scala
package x.y

class CompositeThing
   (components: String*)

object ExampleThing
    extends CompositeThing
          (
                "One",
                "Two",
                "Three",
                "Four")
```

That compiles, but running Scalafmt again on that code however results in this:

```
sbt:test> scalafmt
[info] Formatting 1 Scala sources...
[error] C:\Git\BugReport\src\main\scala\Main.scala:8: error: expected class or object definition (
[error]           (
[error]           ^: C:\Git\BugReport\src\main\scala\Main.scala
[error] (Compile / scalafmt) C:\Git\BugReport\src\main\scala\Main.scala:8: error: expected class or object definition (
[error]           (
[error]           ^: C:\Git\BugReport\src\main\scala\Main.scala
[error] Total time: 0 s, completed 04-Jun-2021 19:13:32
```

Note that, running from within Sbt, I had to touch the file after the first format to force Scalafmt to attempt the second format. I guess Sbt/Scalafmt is being clever by thinking it doesn’t actually need to do the second format if the underlying file it has just formatted hasn’t been touched.

I should be clear that I was still using my local build of `master` at `cdae81bb` when I reported the previous problem. Just in case the `Index 59` issue was specific to my local build, I have just tried formatting the same sample using the official `3.0.0-RC4+38-41d0d30e-SNAPSHOT` build from the command-line. That gives a rather more useful stack trace:

```
C:\Git\BugReport> coursier bootstrap -f "org.scalameta:scalafmt-cli_2.13:3.0.0-RC4+38-41d0d30e-SNAPSHOT" -r sonatype:snapshots --main org.scalafmt.cli.Cli -o scalafmt
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 3.3 KiB (1.2 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 2.8 KiB (1008B / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 3.4 KiB (1.3 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 2.1 KiB (5.1 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 4.3 KiB (9.8 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 114.5 KiB (148.9 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 168.7 KiB (104.5 KiB / s)
https://oss.sonatype.org/content/repositories/snapshots/org/scala...
  100.0% [##########] 1.7 MiB (642.0 KiB / s)
Wrote C:\Git\BugReport\scalafmt
Wrote C:\Git\BugReport\scalafmt.bat

C:\Git\BugReport> .\scalafmt src
org.scalafmt.cli.FailedToFormat: C:\Git\BugReport\src\main\scala\Main.scala
Caused by: java.lang.ArrayIndexOutOfBoundsException: 59
        at org.scalafmt.internal.FormatWriter$FormatLocations.setStats$1(FormatWriter.scala:1130)
        at org.scalafmt.internal.FormatWriter$FormatLocations.setStat$1(FormatWriter.scala:1116)
        at org.scalafmt.internal.FormatWriter$FormatLocations.iter$3(FormatWriter.scala:1167)
        at org.scalafmt.internal.FormatWriter$FormatLocations.org$scalafmt$internal$FormatWriter$FormatLocations$$setTopStats$1(FormatWriter.scala:1170)
        at org.scalafmt.internal.FormatWriter$FormatLocations$$anon$2.apply(FormatWriter.scala:1228)
        at scala.meta.transversers.Traverser.apply(Traverser.scala:4)
        at org.scalafmt.internal.FormatWriter$FormatLocations$$anon$2.apply(FormatWriter.scala:1195)
        at org.scalafmt.internal.FormatWriter$FormatLocations.extraBlankTokens$lzycompute(FormatWriter.scala:1235)
        at org.scalafmt.internal.FormatWriter$FormatLocations.extraBlankTokens(FormatWriter.scala:1089)
        at org.scalafmt.internal.FormatWriter$FormatLocations$Entry.getWhitespace(FormatWriter.scala:282)
        at org.scalafmt.internal.FormatWriter$FormatLocations$Entry.ws$1(FormatWriter.scala:342)
        at org.scalafmt.internal.FormatWriter$FormatLocations$Entry.formatWhitespace(FormatWriter.scala:370)
        at org.scalafmt.internal.FormatWriter.$anonfun$mkString$1(FormatWriter.scala:82)
        at org.scalafmt.internal.FormatWriter.$anonfun$mkString$1$adapted(FormatWriter.scala:48)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1288)
        at org.scalafmt.internal.FormatWriter.mkString(FormatWriter.scala:48)
        at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$6(Scalafmt.scala:129)
        at scala.util.Success.flatMap(Try.scala:258)
        at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$3(Scalafmt.scala:128)
        at scala.meta.parsers.Parsed.fold(Errors.scala:12)
        at scala.meta.parsers.Parsed.fold$(Errors.scala:11)
        at scala.meta.parsers.Parsed$Success.fold(Errors.scala:26)
        at org.scalafmt.Scalafmt$.doFormatOne(Scalafmt.scala:124)
        at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:102)
        at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:71)
        at org.scalafmt.Scalafmt$.format(Scalafmt.scala:52)
        at org.scalafmt.Scalafmt.format(Scalafmt.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.scalafmt.dynamic.ScalafmtReflect.format(ScalafmtReflect.scala:107)
        at org.scalafmt.dynamic.ScalafmtReflectConfig.format(ScalafmtReflectConfig.scala:82)
        at org.scalafmt.dynamic.ScalafmtDynamic.$anonfun$tryFormat$1(ScalafmtDynamic.scala:203)
        at scala.util.Try$.apply(Try.scala:210)
        at org.scalafmt.dynamic.ScalafmtDynamic.org$scalafmt$dynamic$ScalafmtDynamic$$tryFormat(ScalafmtDynamic.scala:188)
        at org.scalafmt.dynamic.ScalafmtDynamic$MySession.format(ScalafmtDynamic.scala:245)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.handleFile(ScalafmtDynamicRunner.scala:110)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$4(ScalafmtDynamicRunner.scala:61)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$4$adapted(ScalafmtDynamicRunner.scala:59)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1288)
        at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:938)
        at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
        at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
        at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
        at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:935)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:152)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
        at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
@@ -2052,8 +2052,10 @@ class FormatOps(
     )(implicit fileLine: FileLine, style: ScalafmtConfig): Seq[Split] = {
       val end = tokens.getLast(tree)
       val slbExpire = nextNonCommentSameLine(end).left
-      val closeOpt = getClosingIfEnclosedInMatching(tree)
-        .map(x => prevNonComment(tokens(x, -1)).left)
+      val closeOpt =
+        if (isTuple(tree)) None
+        else
+          getClosingIfEnclosedInMatching(tree).map(tokens.tokenBefore(_).left)
       def nlPolicy(implicit fileLine: FileLine) =
         decideNewlinesOnlyAfterClose(closeOpt.getOrElse(slbExpire))
       val indentLen =

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1518,11 +1518,8 @@ val (length, width) =
    else
       (secondMarkLength, secondMarkWidth)
 >>>
-Idempotency violated
 val (length, width) =
   if s % 5 == 0 then
-     (hourMarkLength, hourMarkWidth
-  )
+     (hourMarkLength, hourMarkWidth)
   else
-     (secondMarkLength, secondMarkWidth
-  )
+     (secondMarkLength, secondMarkWidth)

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1505,3 +1505,24 @@ end f
 
 /** Some doc comment */
 def g = f("Foo")
+<<< #2561 no forced dangle in tuples
+maxColumn = 40
+newlines {
+  beforeOpenParenDefnSite = true
+  beforeOpenParenCallSite = true
+}
+===
+val (length, width) =
+   if s % 5 == 0 then
+      (hourMarkLength, hourMarkWidth)
+   else
+      (secondMarkLength, secondMarkWidth)
+>>>
+Idempotency violated
+val (length, width) =
+  if s % 5 == 0 then
+     (hourMarkLength, hourMarkWidth
+  )
+  else
+     (secondMarkLength, secondMarkWidth
+  )

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1482,11 +1482,8 @@ val (length, width) =
    else
       (secondMarkLength, secondMarkWidth)
 >>>
-Idempotency violated
 val (length, width) =
   if s % 5 == 0 then
-     (hourMarkLength, hourMarkWidth
-  )
+     (hourMarkLength, hourMarkWidth)
   else
-     (secondMarkLength, secondMarkWidth
-  )
+     (secondMarkLength, secondMarkWidth)

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1469,3 +1469,24 @@ end f
 
 /** Some doc comment */
 def g = f("Foo")
+<<< #2561 no forced dangle in tuples
+maxColumn = 40
+newlines {
+  beforeOpenParenDefnSite = true
+  beforeOpenParenCallSite = true
+}
+===
+val (length, width) =
+   if s % 5 == 0 then
+      (hourMarkLength, hourMarkWidth)
+   else
+      (secondMarkLength, secondMarkWidth)
+>>>
+Idempotency violated
+val (length, width) =
+  if s % 5 == 0 then
+     (hourMarkLength, hourMarkWidth
+  )
+  else
+     (secondMarkLength, secondMarkWidth
+  )

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
@@ -1557,11 +1557,8 @@ val (length, width) =
    else
       (secondMarkLength, secondMarkWidth)
 >>>
-Idempotency violated
 val (length, width) =
   if s % 5 == 0 then
-     (hourMarkLength, hourMarkWidth
-  )
+     (hourMarkLength, hourMarkWidth)
   else
-     (secondMarkLength, secondMarkWidth
-  )
+     (secondMarkLength, secondMarkWidth)

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
@@ -1545,3 +1545,23 @@ end f
 
 /** Some doc comment */
 def g = f("Foo")
+<<< #2561 no forced dangle in tuples
+newlines {
+  beforeOpenParenDefnSite = true
+  beforeOpenParenCallSite = true
+}
+===
+val (length, width) =
+   if s % 5 == 0 then
+      (hourMarkLength, hourMarkWidth)
+   else
+      (secondMarkLength, secondMarkWidth)
+>>>
+Idempotency violated
+val (length, width) =
+  if s % 5 == 0 then
+     (hourMarkLength, hourMarkWidth
+  )
+  else
+     (secondMarkLength, secondMarkWidth
+  )

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
@@ -1664,3 +1664,23 @@ end f
 
 /** Some doc comment */
 def g = f("Foo")
+<<< #2561 no forced dangle in tuples
+newlines {
+  beforeOpenParenDefnSite = true
+  beforeOpenParenCallSite = true
+}
+===
+val (length, width) =
+   if s % 5 == 0 then
+      (hourMarkLength, hourMarkWidth)
+   else
+      (secondMarkLength, secondMarkWidth)
+>>>
+Idempotency violated
+val (length, width) =
+  if s % 5 == 0 then
+     (hourMarkLength, hourMarkWidth
+  )
+  else
+     (secondMarkLength, secondMarkWidth
+  )

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
@@ -1676,11 +1676,8 @@ val (length, width) =
    else
       (secondMarkLength, secondMarkWidth)
 >>>
-Idempotency violated
 val (length, width) =
   if s % 5 == 0 then
-     (hourMarkLength, hourMarkWidth
-  )
+     (hourMarkLength, hourMarkWidth)
   else
-     (secondMarkLength, secondMarkWidth
-  )
+     (secondMarkLength, secondMarkWidth)

